version.hpp.in

```diff
@@ -0,0 +1,9 @@
+#pragma once
```

Comment:
Yeah the `hpp.in` file is the template which allows CMake to generate the `.hpp` file. Normally I wouldn't add generated files to git. I agree that this is a smell!

My reasoning is as follows. There are two ways people are likely to use NamedType:
1. By cloning the repo
2. By downloading a zipped release

In 1., since this is a header only library, people are likely to just clone the repo and start using it, without actually building it. In this case, `version.hpp` will not exist, and they'll get a compilation error from [#include version.hpp](https://github.com/joboccara/NamedType/blob/8efa8cc5acc86da841b34c2837a69e9800385e6e/include/NamedType/named_type.hpp#L6).

In 2., how do you make that release? If you're making it locally or on a build server after having run the build, it would work fine to not commit `version.hpp`. It will have been built and be included in the zip. If however you just ask GitHub to zip up the repo, the file will not exist and will not be included in the zip.

Due to this, I thought it was best to commit this generated file, even though it's usually not the right thing to do.

An alternative to this whole approach would to not have the template at all, and just have the version numbers hardcoded in the file. Then you could either just update those manually before a release, or have a script do that. I understand if you prefer that.

If you want a second opinion, I can ask people on Twitter to have a look at this PR?

Reply:
I get your point, usage is currently easy by just grabbing the code, it would be a shame to lose this, so we'd want a version.hpp file.
But then since we'd have to update the version.hpp file whenever we bump the library version, I think this kills the purpose of the automatic generation, no? If you agree with that, I think we should only keep the manually updated version.hpp file.

Reply:
Yeah I think that makes sense. Currently there's not much extra value in having the generation, except it makes sure you get both the separate number components and the full string correct. By doing it manually there's a theoretic possibility of forgetting one of them.

The approach I used here is more important for bigger projects where you need the version number in other places as well, like package names, documentation etc.

If you want I can change to not using generation. Maybe we could then also stop using a separate version header, and just move the macros into `named_type.hpp`. Let me know what you prefer.

Reply:
Yes I'd be in favour of removing the generation. But I think a separate file makes sense, as this is not the main use case for the library and maybe this is not the code you want to read first when you explore the main header of the library.

Reply:
Ok! I made that change now, updated the commit message, and rebased on master while I was at it. I also added a test to make sure the version number macros and the version string macro are in sync.